packages/core/src/amazonq/lsp/lspController.ts

```diff
@@ -0,0 +1,378 @@
+                    programmingLanguage: {
+                        languageName: chunk.programmingLanguage,
+                    },
+                })
+            } else {
+                resp.push({
+                    text: text,
+                    relativeFilePath: chunk.relativePath ? chunk.relativePath : path.basename(chunk.filePath),
+                })
+            }
+        })
+        return resp
+    }
+
+    async buildIndex() {
+        getLogger().info(`LspController: Starting to build vector index of project`)
+        const start = performance.now()
+        const projPaths = getProjectPaths()
+        projPaths.sort()
+        try {
+            if (projPaths.length === 0) {
+                throw Error('No project')
+            }
+            this._isIndexingInProgress = true
+            const projRoot = projPaths[0]
+            const files = await collectFilesForIndex(
+                projPaths,
+                vscode.workspace.workspaceFolders as CurrentWsFolders,
+                true,
+                CodeWhispererSettings.instance.getMaxIndexSize() * 1024 * 1024
+            )
+            let totalSizeBytes = 0
+            for (let i = 0; i < files.length; i += 1) {
+                totalSizeBytes += files[i].fileSizeBytes
+            }
+            getLogger().info(`LspController: Found ${files.length} files in current project ${getProjectPaths()}`)
+            const resp = await LspClient.instance.indexFiles(
+                files.map(f => f.fileUri.fsPath),
+                projRoot,
+                false
+            )
+            if (resp) {
+                getLogger().debug(`LspController: Finish building vector index of project`)
+                const usage = await LspClient.instance.getLspServerUsage()
+                telemetry.amazonq_indexWorkspace.emit({
+                    duration: performance.now() - start,
+                    result: 'Succeeded',
+                    amazonqIndexFileCount: files.length,
+                    amazonqIndexMemoryUsageInMB: usage ? usage.memoryUsage / (1024 * 1024) : undefined,
+                    amazonqIndexCpuUsagePercentage: usage ? usage.cpuUsage : undefined,
+                    amazonqIndexFileSizeInMB: totalSizeBytes / (1024 * 1024),
+                    credentialStartUrl: AuthUtil.instance.startUrl,
+                })
+            } else {
+                getLogger().error(`LspController: Failed to build vector index of project`)
+                telemetry.amazonq_indexWorkspace.emit({
+                    duration: performance.now() - start,
+                    result: 'Failed',
+                    amazonqIndexFileCount: 0,
+                    amazonqIndexFileSizeInMB: 0,
+                })
+            }
+        } catch (e) {
+            getLogger().error(`LspController: Failed to build vector index of project`)
+            telemetry.amazonq_indexWorkspace.emit({
```

Comment:
can the logic be wrapped in `amazonq_indexWorkspace.run(...)`? that would avoid all of these try/catch cases and redundant `amazonq_indexWorkspace.emit()` calls. And fields like `duration` and `result`, `reason`, and `reasonDesc` would be automatically set.

Reply:
Yes. 

However, I don't understand how this run works with try catch. If I catch the exception and do not throw it, does it set the result to failed with reason?

Reply:
> If I catch the exception and do not throw it, does it set the result to failed with reason?

No. Let run() handle the exception. You can always follow it with a `.catch()` if you don't want the exception to bubble up, but if this is part of a vscode command or webview, then Toolkit automatically handles the exception by showing a user message.